### DISCUSSION TITLE
How to let the decompiler optimize out meaningless functions

### DISCUSSION THREAD

**Discussion by EwTE-H** [NationalSecurityAgency/ghidra#6400] [Q&A]
I'm writing a cspec. The compiler I'm dealing with generates an entrance function and an exit function. The entrance function simply pushes everything, creates a backup of SP, and then returns. The exit function does them in the contradict.
```
; entrance
push qr8
push xr4
mov fp, sp
rt
; exit
mov sp, fp
pop xr4
pop qr8
pop pc
```
Almost every function calls them, making the output messy. What's worse, the decompiler often misunderstands  their usage. For example:
```
call entrance
mov r8, r0
...
mov r0, r8
call exit
```
r0, which is actually a **parameter**, is recognized as the **return value** of `entrance`. The decompiler gives me something like `r8 = entrance(); `. And at return, it gives me `exit(r8); return;`. It's awful.
I wonder whether it is possible to let the decompiler optimize them out.

**Comment by ryanmkurtz** [comment]
Sounds like your cspec needs a couple of `<callfixup>` entries for these functions.  Here is some info on it:

https://spinsel.dev/assets/2020-06-17-ghidra-brainfuck-processor-1/ghidra_docs/compiler_spec/cspec_pcodeinterp.html#idm181

**Reply by EwTE-H** [reply]
But it is for *calls*, not for *returns*. Is there a "returnfixup" ?

**Reply by astrelsky** [reply]
> But it is for *calls*, not for *returns*. Is there a "returnfixup" ?

Have the call fixup do something like mov r0, r8 then it'll just be meaningless.
